src/shared/completion/use-set-form-completion-mutation.test.js

```diff
@@ -0,0 +1,249 @@
+        const queryCache = new QueryCache()
+        const onDataValueSuccess = jest.fn()
+        const dataValuesResolver = jest.fn(() => new Promise(() => null))
+
+        // first we need to load the data value set so we have a loading state
+        const { result: dataValueSet, waitFor: waitForDataValueSet } =
+            renderHook(
+                () => useDataValueSet({ onSuccess: onDataValueSuccess }),
+                {
+                    wrapper: ({ children }) => (
+                        <Wrapper
+                            queryClientOptions={{ queryCache }}
+                            dataForCustomProvider={{
+                                dataValues: dataValuesResolver,
+                            }}
+                        >
+                            {children}
+                        </Wrapper>
+                    ),
+                }
+            )
+
+        // Make sure the request is on-going
+        await waitForDataValueSet(() => {
+            expect(dataValuesResolver).toHaveBeenCalledTimes(1)
+            expect(onDataValueSuccess).toHaveBeenCalledTimes(0)
+            return dataValueSet.current.isLoading
+        })
+
+        const { result: setFormCompletion, waitFor: waitForSetFormCompletion } =
+            renderHook(() => useSetFormCompletionMutation(), {
+                wrapper: ({ children }) => (
+                    <Wrapper
+                        queryClientOptions={{ queryCache }}
+                        dataForCustomProvider={{
+                            // never resolving
+                            'dataEntry/dataSetCompletion': () =>
+                                new Promise(() => null),
+                        }}
+                    >
+                        {children}
+                    </Wrapper>
+                ),
+            })
+
+        expect(dataValueSet.current.isLoading).toBe(true)
+
+        // Mutate a current data value
+        act(() => setFormCompletion.current.mutate({ completed: true }))
+
+        await waitForSetFormCompletion(() => {
+            expect(setFormCompletion.current.isLoading).toBe(true)
+        })
+
+        // Ensure that the on-going data value set request is cancelled
+        await waitForDataValueSet(() => {
+            // We're still waiting for the form completion response, so it's
+            // guaranteed that the data value set query hasn't been retriggered
+            // yet
+            expect(setFormCompletion.current.isLoading).toBe(true)
+            expect(dataValuesResolver).toHaveBeenCalledTimes(1)
+
+            // useDataValueSet is not loading but never has finished, which
+            // verifies that the active query has been cancelled
+            expect(dataValueSet.current.isLoading).toBe(false)
+            expect(onDataValueSuccess).toHaveBeenCalledTimes(0)
+        })
+    })
+
+    it('should optimistically update the completion state', async () => {
```

Comment:
I don't get some things with this test. Why mock the a datavalueset request? I would assume that we do not need to test that this mutation is cancelled as it's covered in the previous test? I would think that the test (from the description) would just be testing that the completion state in the queryCache goes from false to true?

Reply:
True. I've removed the unnecessary code